docker/reference/regexp.go

```diff
@@ -1,143 +1,168 @@
+	anchoredDigestRegexp = match(anchored(digestStr))
+
+	// The string counterpart for NameRegexp.
+	nameStr = expression(
+		optional(domainStr, literal(`/`)),
+		nameComponentStr,
+		optional(repeated(literal(`/`), nameComponentStr)))
 
 	// NameRegexp is the format for the name component of references. The
 	// regexp has capturing groups for the domain and name part omitting
 	// the separating forward slash from either.
-	NameRegexp = expression(
-		optional(DomainRegexp, literal(`/`)),
-		nameComponentRegexp,
-		optional(repeated(literal(`/`), nameComponentRegexp)))
+	NameRegexp = match(nameStr)
 
 	// anchoredNameRegexp is used to parse a name value, capturing the
 	// domain and trailing components.
-	anchoredNameRegexp = anchored(
-		optional(capture(DomainRegexp), literal(`/`)),
-		capture(nameComponentRegexp,
-			optional(repeated(literal(`/`), nameComponentRegexp))))
+	anchoredNameRegexp = match(anchored(
+		optional(capture(domainStr), literal(`/`)),
+		capture(nameComponentStr,
+			optional(repeated(literal(`/`), nameComponentStr)))))
+
+	// The string counterpart for ReferenceRegexp
+	referenceStr = anchored(capture(nameStr),
+		optional(literal(":"), capture(tagStr)),
+		optional(literal("@"), capture(digestStr)))
 
 	// ReferenceRegexp is the full supported format of a reference. The regexp
 	// is anchored and has capturing groups for name, tag, and digest
 	// components.
-	ReferenceRegexp = anchored(capture(NameRegexp),
-		optional(literal(":"), capture(TagRegexp)),
-		optional(literal("@"), capture(DigestRegexp)))
+	ReferenceRegexp = match(referenceStr)
 
 	// IdentifierRegexp is the format for string identifier used as a
 	// content addressable identifier using sha256. These identifiers
 	// are like digests without the algorithm, since sha256 is used.
-	IdentifierRegexp = match(`([a-f0-9]{64})`)
+	IdentifierRegexp = match(identifierStr)
 
 	// ShortIdentifierRegexp is the format used to represent a prefix
 	// of an identifier. A prefix may be used to match a sha256 identifier
 	// within a list of trusted identifiers.
-	ShortIdentifierRegexp = match(`([a-f0-9]{6,64})`)
+	ShortIdentifierRegexp = match(shortIdentifierStr)
 
 	// anchoredIdentifierRegexp is used to check or match an
 	// identifier value, anchored at start and end of string.
-	anchoredIdentifierRegexp = anchored(IdentifierRegexp)
+	anchoredIdentifierRegexp = match(anchored(identifierStr))
 
 	// anchoredShortIdentifierRegexp is used to check if a value
 	// is a possible identifier prefix, anchored at start and end
 	// of string.
-	anchoredShortIdentifierRegexp = anchored(ShortIdentifierRegexp)
+	anchoredShortIdentifierRegexp = match(anchored(shortIdentifierStr))
 )
 
 // match compiles the string to a regular expression.
 var match = regexp.MustCompile
 
 // literal compiles s into a literal regular expression, escaping any regexp
 // reserved characters.
-func literal(s string) *regexp.Regexp {
+func literal(s string) string {
```

Comment:
AFAICS this could be basically eliminated (by making it `return regexp.QuoteMeta` or just hard-coding the result), avoiding another bunch of compiles.